WalletWasabi/Wallets/P2pBlockProvider.cs

```diff
@@ -0,0 +1,341 @@
+							}
+
+							// Validate block
+							if (!block.Check())
+							{
+								Logger.LogInfo($"Disconnected node: {node.RemoteSocketAddress}, because invalid block received.");
+								node.DisconnectAsync("Invalid block received.");
+								continue;
+							}
+
+							if (Nodes.ConnectedNodes.Count > 1) // To minimize risking missing unconfirmed transactions.
+							{
+								Logger.LogInfo($"Disconnected node: {node.RemoteSocketAddress}. Block downloaded: {block.GetHash()}.");
+								node.DisconnectAsync("Thank you!");
+							}
+
+							await NodeTimeoutsAsync(false);
+						}
+						catch (Exception ex) when (ex is OperationCanceledException || ex is TaskCanceledException || ex is TimeoutException)
+						{
+							Logger.LogInfo($"Disconnected node: {node.RemoteSocketAddress}, because block download took too long.");
+
+							await NodeTimeoutsAsync(true);
+
+							node.DisconnectAsync("Block download took too long.");
+							continue;
+						}
+						catch (Exception ex)
+						{
+							Logger.LogDebug(ex);
+							Logger.LogInfo($"Disconnected node: {node.RemoteSocketAddress}, because block download failed: {ex.Message}.");
+							node.DisconnectAsync("Block download failed.");
+							continue;
+						}
+
+						break; // If got this far, then we have the block and it's valid. Break.
+					}
+					catch (Exception ex)
+					{
+						Logger.LogDebug(ex);
+					}
+				}
+			}
+			finally
+			{
+				DownloadingBlockChanged?.Invoke(null, false);
+			}
+
+			return block;
+		}
+
+		private async Task<Block> TryDownloadBlockFromLocalNodeAsync(uint256 hash, CancellationToken cancellationToken)
+		{
+			if (CoreNode?.RpcClient is null)
+			{
+				try
+				{
+					if (LocalBitcoinCoreNode is null || (!LocalBitcoinCoreNode.IsConnected && Network != Network.RegTest)) // If RegTest then we're already connected do not try again.
+					{
+						DisconnectDisposeNullLocalBitcoinCoreNode();
+						using var handshakeTimeout = CancellationTokenSource.CreateLinkedTokenSource(cancellationToken);
+						handshakeTimeout.CancelAfter(TimeSpan.FromSeconds(10));
+						var nodeConnectionParameters = new NodeConnectionParameters()
+						{
+							ConnectCancellation = handshakeTimeout.Token,
+							IsRelay = false,
+							UserAgent = $"/Wasabi:{Constants.ClientVersion.ToString()}/"
+						};
+
+						// If an onion was added must try to use Tor.
+						// onlyForOnionHosts should connect to it if it's an onion endpoint automatically and non-Tor endpoints through clearnet/localhost
+						if (Synchronizer.WasabiClient.TorClient.IsTorUsed)
+						{
+							nodeConnectionParameters.TemplateBehaviors.Add(new SocksSettingsBehavior(Synchronizer.WasabiClient.TorClient.TorSocks5EndPoint, onlyForOnionHosts: true, networkCredential: null, streamIsolation: false));
+						}
+
+						var localEndPoint = ServiceConfiguration.BitcoinCoreEndPoint;
+						var localNode = await Node.ConnectAsync(Network, localEndPoint, nodeConnectionParameters);
```

Comment:
.ConfigureAwait(false);

Reply:
This is as it is was, I didn't change it. I will do it anyway.

Reply:
Done.